### PR TITLE
CE-1304 Enable targeting a SiteWideMessage to Power Users

### PR DESCRIPTION
**Pull request by adamkarminski**
[@see CE-1281](https://wikia-inc.atlassian.net/browse/CE-1281)

This PR enables targeting of SiteWideMessages to all or specific power users.

**IMPORTANT:** Unfortunanately, we can only fix one thing at once. This PR uses the existing, spaghetti architecture of SWM and tries to keep up to standards as much as possible.

Ping @Wikia/community-engineering

### CODE REVIEW

**Comment by lgarczewski** [comment]
LGTM. Wait for CI to complete and then good to merge.

File: extensions/wikia/SiteWideMessages/templates/editor.tmpl.php [hunk 1]
```diff
@@ -163,6 +163,31 @@
...
+						<tr>
+							<td>
+								<input name="mSendModeUsers" id="mSendModeUsersPU" type="radio" value="POWERUSER"<?= $formData['sendModeUsers'] == 'POWERUSER' ? ' checked="checked"' : ''?>/>
+							</td>
+							<td>
+								<label for="mSendModeUsersPU"><?= wfMessage( 'swm-label-mode-users-poweruser' )->escaped() ?></label>
+							</td>
+							<td>
+								<?php
+									foreach	( $formData['powerUserTypes'] as $powerUserType ) {
```
Comment (Grunny) [inline extensions/wikia/SiteWideMessages/templates/editor.tmpl.php]: Where is `powerUserTypes` set on the form data? I don't see these checkboxes being added only the single send mode option.

File: extensions/wikia/SiteWideMessages/templates/editor.tmpl.php [hunk 2]
```diff
@@ -163,6 +163,34 @@
...
+										$label = '<label for="mPowerUserType_' . Sanitizer::encodeAttribute( $powerUserType ) . '">';
+											$label .= Sanitizer::escapeHtmlAllowEntities( $powerUserType ) . '</label><br>';
+										$html = '<input type="checkbox" name="mPowerUserType[]"';
+											$html .= 'id="mPowerUserType_' . Sanitizer::encodeAttribute( $powerUserType ) . '"';
+											$html .= 'value="' . Sanitizer::encodeAttribute( $powerUserType ) . '"';
+										if ( isset( $formData['mPowerUserType'] ) && in_array( $powerUserType, $formData['mPowerUserType'] ) ) {
+											$html .= ' checked="checked"';
+										}
+										$html .= '>';
+										echo $html;
```
Comment (Grunny) [inline extensions/wikia/SiteWideMessages/templates/editor.tmpl.php]: In general, it's always better to wrap a checkbox in the label so there isn't an area between the checkbox and the text that can't be clicked.

File: extensions/wikia/SiteWideMessages/templates/editor.tmpl.php [hunk 3]
```diff
@@ -410,8 +446,9 @@ function grayOut( e ) {
 	$( '#mSendModeWikisA' ).add( '#mSendModeWikisH' ).add( '#mSendModeWikisC' )
 		.add( '#mSendModeWikisW' ).add( '#mSendModeWikisM' ).add( '#mSendModeWikisD' )
 		.add( '#mSendModeUsersA' ).add( '#mSendModeUsersC' ).add( '#mSendModeUsersG' )
-		.add( '#mSendModeUsersU' ).add( '#mSendModeUsersM' ).add( '#mSendModeUsersIP' )
-		.add( '#mSendModeUsersReg' ).add( '#mSendModeUsersEC' ).bind( 'click', grayOut );
+		.add( '#mSendModeUsersPU' ).add( '#mSendModeUsersU' ).add( '#mSendModeUsersM' )
+		.add( '#mSendModeUsersIP' ).add( '#mSendModeUsersReg' ).add( '#mSendModeUsersEC' )
+		.bind( 'click', grayOut );
```
Comment (Grunny) [inline extensions/wikia/SiteWideMessages/templates/editor.tmpl.php]: Not part of this change, but if you're in there, in jQuery 1.7+ `.on` is preferred over `.bind`.

File: extensions/wikia/SiteWideMessages/SpecialSiteWideMessages.i18n.php [hunk 4]
```diff
@@ -49,6 +49,8 @@
 	'swm-label-mode-users-active'		=> 'Active users',
 	'swm-label-mode-users-group'		=> 'Users belonging to the group',
 	'swm-label-mode-users-group-hint'	=> '<i>Pick a group from drop down or write name by hand to overwrite drop down selection.</i>',
+	'swm-label-mode-users-poweruser' => 'Power users of a type',
+	'swm-label-mode-users-poweruser-hint' => 'Pick types of power users from a list. The message will be sent to All Wikias.',
```
Comment (Grunny) [inline extensions/wikia/SiteWideMessages/SpecialSiteWideMessages.i18n.php]: Can we get QQQ docs for these? This extension is translated by TWN.

Comment (lukaszkonieczny) [inline extensions/wikia/SiteWideMessages/SpecialSiteWideMessages.i18n.php]: Did you add qqq description?

File: includes/wikia/tasks/Tasks/SiteWideMessagesTask.class.php [hunk 5]
```diff
@@ -157,6 +161,58 @@ private function sendMessageToGroup($params) {
...
+		 * Get IDs of users with the specified properties
+		 */
+		$userIds = ( new \WikiaSQL() )
+			->SELECT()->DISTINCT( 'up_user' )
+			->FROM( 'user_properties' )
+			->WHERE( 'up_property' )->IN( $powerUsersTypesArr )
+			->AND_( 'up_value' )->EQUAL_TO( 1 )
+			->runLoop( $DB, function( &$userIds, $row ) {
+				$userIds[] = $row->up_user;
+			});
```
Comment (Grunny) [inline includes/wikia/tasks/Tasks/SiteWideMessagesTask.class.php]: `});` -> `} );`

File: includes/wikia/tasks/Tasks/SiteWideMessagesTask.class.php [hunk 6]
```diff
@@ -157,6 +161,58 @@ private function sendMessageToGroup($params) {
...
+		/**
+		 * Create a messages_status record for each ID
+		 */
+		foreach ( $userIds as $userId ) {
+			$sqlValues[] = [ 0, $userId, $params[ 'messageId' ], MSG_STATUS_UNSEEN ];
+		}
+
+		$this->info('add records about new message to users', [
+			'num_users' => count( $sqlValues )
+		]);
```
Comment (Grunny) [inline includes/wikia/tasks/Tasks/SiteWideMessagesTask.class.php]: `});` -> `} );`

File: extensions/wikia/SiteWideMessages/SpecialSiteWideMessages.i18n.php (hunk 4 above)
Comment (adamkarminski) [inline extensions/wikia/SiteWideMessages/SpecialSiteWideMessages.i18n.php]: Done :+1:

**Comment by lukaszkonieczny** [comment]
lgtm
